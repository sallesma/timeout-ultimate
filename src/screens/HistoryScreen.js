import React, { useEffect, useState } from 'react';
import { StatusBar } from 'expo-status-bar';
import { ScrollView, StyleSheet, Text, View } from 'react-native';

import { loadQuizzResults } from '../utils/history';
import theme from '../utils/theme.js';
import I18n from '../utils/i18n';

function formatDate(date) {
  const d = new Date(date),
    year = d.getFullYear();
  let month = '' + (d.getMonth() + 1),
    day = '' + d.getDate();

  if (month.length < 2) month = '0' + month;
  if (day.length < 2) day = '0' + day;

  return [day, month, year].join('/');
}

export default () => {
  const [results, setResults] = useState([]);

  useEffect(() => {
    (async function loadResults() {
      const loadedResults = await loadQuizzResults();
      setResults(loadedResults);
    })();
  }, []);

  const orderedResults = results.slice(0).reverse();

  let totalQuestions = 0,
    totalRightAnswers = 0;
  orderedResults.forEach((result) => {
    totalQuestions += result.quizzLength;
    totalRightAnswers += result.rightAnswersCount;
  });

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.headerText}>{I18n.t('historyScreen.questions', { count: totalQuestions })}</Text>
        <Text style={styles.headerText}>{I18n.t('historyScreen.rightAnwsers', { count: totalRightAnswers })}</Text>
        {results.length > 0 && (
          <Text style={styles.headerText}>
            {I18n.t('historyScreen.percentage', { count: Math.round((totalRightAnswers / totalQuestions) * 100) })}
          </Text>
        )}
      </View>
      {results.length === 0 && <Text style={styles.incentive}>{I18n.t('historyScreen.empty')}</Text>}
      <ScrollView>
        {orderedResults.map((result, index) => (
          <Text key={index}>
            {formatDate(new Date(result.createdAt))} - 
            {I18n.t('historyScreen.result', {
              count: result.rightAnswersCount,
              total: result.quizzLength,
              percentage: ((result.rightAnswersCount * 100) / result.quizzLength).toFixed(),
            })}
          </Text>
        ))}
      </ScrollView>
      <StatusBar style="auto" />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 16,
  },
  header: {
    paddingBottom: 16,
    marginBottom: 8,
    borderBottomWidth: 1,
    borderBottomColor: theme.MAIN_COLOR,
  },
  headerText: {
    fontSize: theme.FONT_SIZE_L,
  },
  incentive: {
    marginTop: 32,
    textAlign: 'center',
    fontSize: theme.FONT_SIZE_M,
  },
});
