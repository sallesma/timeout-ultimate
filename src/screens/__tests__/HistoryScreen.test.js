import React from 'react';
import { render, waitFor } from '@testing-library/react-native';

import HistoryScreen from '../HistoryScreen';

describe('<HistoryScreen />', () => {
  it('renders correctly', async () => {
    const { toJSON } = await waitFor(() => render(<HistoryScreen />));

    expect(toJSON()).toMatchSnapshot();
  });
});
