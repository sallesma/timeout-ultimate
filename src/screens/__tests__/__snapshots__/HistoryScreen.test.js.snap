// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HistoryScreen /> renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      {
        "borderBottomColor": "#2a9d9d",
        "borderBottomWidth": 1,
        "marginBottom": 8,
        "paddingBottom": 16,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 20,
        }
      }
    >
      Questions: 0
    </Text>
    <Text
      style={
        {
          "fontSize": 20,
        }
      }
    >
      Correct answers: 0
    </Text>
  </View>
  <Text
    style={
      {
        "fontSize": 16,
        "marginTop": 32,
        "textAlign": "center",
      }
    }
  >
    There are no results because you haven't taken the quiz yet!
  </Text>
  <RCTScrollView>
    <View />
  </RCTScrollView>
</View>
`;
